### PR TITLE
custom domain support 

### PR DESCRIPTION
**Pull request by v1xingyue**
This help gateway  support the gateway custom domain name by [autocert](golang.org/x/crypto/acme/autocert). 

At first user should have the web3 deployed contract like : 

http://0xe15406f9092cb0cb77191fa3cbee92df07c9cc8b.80001.w3link.io/getMessage

Then add the Cname record as 

yourname.example.com ->  with value 0xe15406f9092cb0cb77191fa3cbee92df07c9cc8b.80001.w3link.io . 

At last visit  https://yourname.example.com .

### CODE REVIEW

File: cmd/server/autossl.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+	"context"
+
+	"golang.org/x/crypto/acme/autocert"
+)
+
+var certManager = autocert.Manager{
+	Prompt: autocert.AcceptTOS,
+	// HostPolicy: autocert.HostWhitelist(domain),
+	HostPolicy: func(ctx context.Context, host string) error {
+		return nil
```
Comment (qzhodl) [inline cmd/server/autossl.go:13]: does it mean that we will try to generate an SSL certificate for any requesting domain?

Comment (v1xingyue) [reply cmd/server/autossl.go:13]: Yes, That's true! Also we can add limit with this function.

File: cmd/server/protocol.go [hunk 2]
```diff
@@ -16,10 +16,19 @@ import (
...
-
 func handle(w http.ResponseWriter, req *http.Request) {
+
 	h := req.Host
 
+	if cname, err := net.LookupCNAME(h); err == nil {
+		log.Infof("cname is ---> %s", cname)
+		if strings.HasSuffix(cname, ".") {
+			h = cname[:len(cname)-1]
+			w.Header().Set("We3-CNAME", cname)
```
Comment (syntrust) [inline cmd/server/protocol.go:27]: We3 -> Web3?

File: cmd/server/main.go [hunk 3]
```diff
@@ -248,9 +250,24 @@ func main() {
 			return
 		}
 	} else {
-		err := http.ListenAndServeTLS(":"+config.ServerPort, config.CertificateFile, config.KeyFile, nil)
-		if err != nil {
+		server := &http.Server{
```
Comment (syntrust) [inline cmd/server/main.go:253]: How should CertificateFile and KeyFile be configured to get us here?

File: cmd/server/protocol.go (hunk 2 above)
Comment (v1xingyue) [reply cmd/server/protocol.go]: Ya. That should be Web3.

File: cmd/server/main.go (hunk 3 above)
Comment (v1xingyue) [reply cmd/server/main.go:253]: This use autocert lib manage the ssl site the 443 port will be used. Also we may add the white list to the ssl certificate generation.

Comment (qzhodl) [reply cmd/server/main.go:253]: I think @syntrust was asking how to config the toml file so that the gateway can go to the "else" case

Comment (v1xingyue) [reply cmd/server/main.go:253]: That's very simple. Just add anyting not empty  with value `CertificateFile` and `KeyFile`. 
Then will go to the else autossl  mode.
Another thing, cause cert does not come from toml config file, maybe we can add another value like 
```yaml
httpMode: true
```
Then gateway will run as http mode or https mode.

Comment (qzhodl) [reply cmd/server/main.go:253]: Yes, I think that sounds much better

Comment (v1xingyue) [reply cmd/server/main.go:254]: Done as last commit .

File: cmd/server/autossl.go (hunk 1 above)
Comment (qzhodl) [reply cmd/server/autossl.go:13]: We ran the PR code on a test machine and configured it to web3gateway.dev. But it seems that https://0xff970a61a04b1ca14834a43f5de4533ebddb5cc8.arb1.web3gateway.dev/name?returns=(string) didn't work, but https://0xff970a61a04b1ca14834a43f5de4533ebddb5cc8.arb1.w3link.io/name?returns=(string) works. Any ideas?

Comment (v1xingyue) [reply cmd/server/autossl.go:13]: My iPhone works both same image

Comment (qizhou) [reply cmd/server/autossl.go:13]: > We ran the PR code on a test machine and configured it to web3gateway.dev. But it seems that https://0xff970a61a04b1ca14834a43f5de4533ebddb5cc8.arb1.web3gateway.dev/name?returns=(string) didn't work, but https://0xff970a61a04b1ca14834a43f5de4533ebddb5cc8.arb1.w3link.io/name?returns=(string) works. Any ideas?

Both links also work for me.

File: cmd/server/autossl.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		Email: config.AutoCertEmail,
+	}
+)
+
+func getCertFromPath(domain, path string) (*tls.Certificate, error) {
+	var (
+		findCert *tls.Certificate = nil
+		findErr                   = fmt.Errorf("certificate file %s invalid", path)
+	)
+	if data, err := os.ReadFile(path); err == nil {
```
Comment (syntrust) [inline cmd/server/autossl.go:51]: The logic seems to work good.  The code may look better to use the idiomatic`
	if err != nil {
		return err
	} ` ?

File: cmd/server/autossl.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+			return findCert, findErr
+		} else {
+			cache.Delete(context.Background(), domainSysCertPath)
+		}
+	}
+
+	if config.SystemCertDir != "" {
+		if stat, err := os.Stat(config.SystemCertDir); err == nil && stat.IsDir() {
+			filepath.WalkDir(config.SystemCertDir, func(path string, d os.DirEntry, err error) error {
+				if err == nil {
```
Comment (syntrust) [inline cmd/server/autossl.go:105]: let's flatten if else here too

**Review by qzhodl — APPROVED** [review]

**Review by syntrust — APPROVED** [review]

**Pull request merged**